state_transfer/endpoint/ingress/ingress.go

```diff
@@ -0,0 +1,191 @@
+
+func (i *IngressEndpoint) NamespacedName() types.NamespacedName {
+	return i.namespacedName
+}
+
+func (i *IngressEndpoint) Labels() map[string]string {
+	return i.labels
+}
+
+func (i *IngressEndpoint) IsHealthy(c client.Client) (bool, error) {
+	ing := &networkingv1.Ingress{}
+	err := c.Get(context.TODO(), i.NamespacedName(), ing)
+	if err != nil {
+		return false, err
+	}
+	if len(ing.Spec.Rules) > 0 && ing.Spec.Rules[0].Host == "" {
+		return false, fmt.Errorf("hostname not set for ingress: %s", ing)
+	}
+
+	if len(ing.Status.LoadBalancer.Ingress) > 0 && ing.Status.LoadBalancer.Ingress[0].Hostname != "" {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (i *IngressEndpoint) createIngressService(c client.Client) error {
+	serviceSelector := i.Labels()
+
+	service := corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      i.NamespacedName().Name,
+			Namespace: i.NamespacedName().Namespace,
+			Labels:    i.Labels(),
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Name:     i.NamespacedName().Name,
+					Protocol: corev1.ProtocolTCP,
+					Port:     i.Port(),
+					TargetPort: intstr.IntOrString{
+						Type:   intstr.Int,
+						IntVal: i.Port()},
+				},
+			},
+			Selector: serviceSelector,
+			Type:     corev1.ServiceTypeNodePort,
+		},
+	}
+	// TODO: consider patching an existing object if it already exists
+	err := c.Create(context.TODO(), &service, &client.CreateOptions{})
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		return err
+	}
+
+	return nil
+}
+
+func (i *IngressEndpoint) createIngress(c client.Client) error {
+	pathType := networkingv1.PathTypePrefix
+	ing := networkingv1.Ingress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      i.NamespacedName().Name,
+			Namespace: i.NamespacedName().Namespace,
+			Labels:    i.Labels(),
+			Annotations: map[string]string{
+				"nginx.ingress.kubernetes.io/ssl-passthrough": "true",
+			},
+		},
+		Spec: networkingv1.IngressSpec{
+			Rules: []networkingv1.IngressRule{{
+				Host: i.hostname,
+				IngressRuleValue: networkingv1.IngressRuleValue{
+					HTTP: &networkingv1.HTTPIngressRuleValue{
+						Paths: []networkingv1.HTTPIngressPath{{
+							Backend: networkingv1.IngressBackend{
+								Service: &networkingv1.IngressServiceBackend{
+									Name: i.namespacedName.Name,
+									Port: networkingv1.ServiceBackendPort{
+										Number: i.port,
+									},
+								},
+							},
+							Path:     "/",
+							PathType: &pathType,
+						}},
+					},
+				},
+			}},
+		},
+	}
+
+	err := c.Create(context.TODO(), &ing, &client.CreateOptions{})
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		return err
+	}
+	return nil
+}
+
+func NewEndpoint(namespacedName types.NamespacedName, labels map[string]string) endpoint.Endpoint {
+	i := &IngressEndpoint{
+		namespacedName: namespacedName,
+		labels:         labels,
+		port:           6443,
+		// TODO: bring the subdomain as a param
+		hostname: namespacedName.Name + ".crane.dev",
```

Comment:
Is the using of the `subdomain` something you intend to include in this PR?

Reply:
yes, will push an update. Gimme a few mins, good catch